### ISSUE TITLE
Suggestion: Rope limit

### ISSUE THREAD

**Issue by deinemudda32** [Facepunch/garrysmod-requests#555]
I yesterday realised, that Garry's Mod Sandbox doesn't have a Ropelimit.
I would like to see it added, as my server often gets spammed with thousand of ropes by random minges.

**Comment by Noiwex** [comment]
Then don't give them rope tool you moron.

**Comment by deinemudda32** [comment]
But I want to, you moron.

**Comment by Noiwex** [comment]
Then you're horrible server admin.

**Comment by deinemudda32** [comment]
For not restricting a tool that everyone wants.
You're right! Give this guy a price!

**Comment by sirsavary** [comment]
+1

**Comment by Kefta** [comment]
@Noiwex It doesn't matter; there's a limit for all other things in Sandbox.

**Comment by Noiwex** [comment]
-1
Ropes do not hurt

**Comment by Kefta** [comment]
@Noiwex Yes they do, it's called spam. A limit convar is always in place for stuff that can be spammed/abused.

**Comment by xaviergmail** [comment]
Noiwex
Stop post

**Comment by kklouzal** [comment]
:+1: Rope limits, limits for all constraints.

**Comment by Noiwex** [comment]
Especially weld

**Comment by mcd1992** [comment]
Welds aren't visible noi.

**Comment by thegrb93** [comment]
Welds do consume resources though.

**Comment by immibis** [comment]
+1 for limits on constraints - it makes perfect sense.

**Comment by AiredaleExtreme** [comment]
+1 I like to host a dedicated server for friends and myself to play in. The computer I use isn't very great. The server crashes when more than ten ropes are used. So, being able to set a limit for constraints would be very beneficial to my server, and probably to other servers that running on older machines as well.

**Comment by AiredaleExtreme** [comment]
By the way, don't be that guy who's like, 'Well, why don't you just upgrade your equipment. =P'.

I'm low on cash, but I will do that as soon as I can.

In any case, too many restraints can lag/crash any server, not just ones on old machines like mine.

**Comment by Noiwex** [comment]
Can you make default limit of 1 rope? Pretty sure it should be enough.

**Comment by thegrb93** [comment]
I've had people devastate my server with dupes that contain 1000s of ropes. It is a problem so stop joking around.

**Comment by robotboy655** [comment]
For the very same reason as https://github.com/Facepunch/garrysmod-requests/issues/21 this would not be possible to do properly, with duplicator support in mind. It can be done, but duplicator will be able to bypass the limit. If we limit duplicator, all tools that use the constraint library (with different limits) would be affected.

**Comment by thegrb93** [comment]
We can't just create a playerspawnconstraint hook which checks and increments the constraint count?

**Comment by robotboy655** [comment]
And what's the point in that? The problem is not in counting the constraints, the problem is in blocking them in duplications.

**Comment by thegrb93** [comment]
Have the duplicator check the playerspawnconstraint hook just like how it checks the playerspawnprop hook.

**Comment by robotboy655** [comment]
There cannot be a `PlayerSpawnConstraint` hook because not every constraint function has a player argument.

That being said, I have added `sbox_maxconstraints` and `sbox_maxropeconstraints`, both default to 1000.

`sbox_maxconstraints` is for non rope constraints only, **not** rope and non-rope combined.

Nocollide is considered a non rope constraint now for the sake of limits.

As I said before, the duplicator support is not perfect. I managed to mostly work around the limitations, but if you hit one of the limits when spawning duplications, both types of constraints will be blocked.

Also I fixed a bug where duplicated constraints would not be able to be cleaned up via Utilities > Admin or User Cleanup
